language/diem-vm/parallel-executor/src/executor.rs

```diff
@@ -7,16 +7,53 @@ use crate::{
     scheduler::Scheduler,
     task::{ExecutionStatus, ExecutorTask, ReadWriteSetInferencer, Transaction, TransactionOutput},
 };
-use anyhow::Result as AResult;
-use mvhashmap::MVHashMap;
+use anyhow::{bail, Result as AResult};
+use mvhashmap::{MVHashMap, Version};
 use num_cpus;
 use rayon::{prelude::*, scope};
 use std::{
     cmp::{max, min},
+    hash::Hash,
     marker::PhantomData,
-    sync::Arc,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
 };
 
+pub struct MVHashMapView<'a, K, V> {
+    map: &'a MVHashMap<K, V>,
+    version: Version,
+    scheduler: &'a Scheduler,
+    has_unexpected_read: AtomicBool,
+}
+
+impl<'a, K: Hash + Clone + Eq, V> MVHashMapView<'a, K, V> {
+    pub fn read(&self, key: &K) -> AResult<Option<&V>> {
+        match self.map.read(key, self.version) {
+            Ok(v) => Ok(Some(v)),
+            Err(None) => Ok(None),
+            Err(Some(dep_idx)) => {
+                // Mark transaction `idx` to be dependent on `dep_idx`.
```

Comment:
I'm not sure if the comment or the logic is wrong, but looking at the code for `add_dependency`, it seems like it's doing the opposite of what the comment says (i.e., marking `self.version` as a dependency of `dep_idx`.